### DISCUSSION TITLE
How to accomplish default props with an MDX component?

### DISCUSSION THREAD

**Discussion by mattvague** [mdx-js/mdx#2059] [Q&A] [answered]
Say you have an MDX file like the following which uses a prop:

```mdx
# My Heading

My prop: {props.a}
```

What would be the "MDX way" to set a default value for that prop? 

In a JSX file, I would do one of the following:

```jsx
export function MyComp({ a = "value", ...props }) {
  return <div>My prop: {props.a}</div>
}

export function MyComp(props = { a: 'value' }) {
  return <div>My prop: {props.a}</div>
}

export function MyComp(props) {
  return <div>My prop: {props.a}</div>
}
MyComp.defaultProps = { a: 'value' }
```

**I've experimented quite a bit and searched around but am still at a loss for how to do that in MDX (if it's even possible).**

---

If it's not currently possible, would it make sense to create a plugin that allows you do something like this?

```mdx
# My Heading

My prop: {prop.a}

export function getDefaultProps(props) {
   return { ...props, a: 'value' }
}
```

I'm imagining I that it would automatically get called inside of MDXContent like this or something:

```jsx
export function MDXContent(initialProps) {
  const props = getDefaultProps(initialProps)

  return <>
    <h1>My Heading</h1>
    My prop: {props.a}
  </>
}
```

It might even be something that should be supported natively in MDX, would you accept a PR to add it?

**Comment by ChristianMurphy** [comment, marked as answer]
So taking a step back, how are you using MDX in the broader context of your application.
Generally, I've seen and used MDX to generate pages or content items.
But it sounds like you are interested in using it more a generic component builder? (Is that accurate?)

---

It's also worth nothing that there are a few options which could work with MDX today, without modifications.

## first option - a get prop utility

If you don't want to customize MDX, and want the props to apply to all usages automatically, something like:

```jsx
export function getProp(props, name) {
  const defaultProps = {
    // put defaults here
  }
  return {...defaultProps, ...props}[name]
}

# My Heading

My prop: {getProp(props, 'a')}
```

## second option - nullish coalesing

If a prop is only used in one place:

```jsx
# My Heading

My prop: {props.a ?? 'test'}
```

or

```jsx
# My Heading

My prop: {props.a || 'test'}
```

can be used

## third option - create an ESAST plugin

ESAST plugins allow the final javascript generated to be modified via an AST.
https://unifiedjs.com/learn/ provides an into to ASTs.
And https://github.com/remcohaszing/recma-nextjs-static-props provides an example of working with props in ESAST.

---

> It might even be something that should be supported natively in MDX, would you accept a PR to add it?

I'm open to discussing it.
My initial reaction/thought is that having a magic function to handle this specific use case sounds like it could cause as much if not more confusion, than it would solve a problem and add clarity.
But open to hearing thoughts otherwise.

**Reply by mattvague** [reply]
>So taking a step back, how are you using MDX in the broader context of your application.
Generally, I've seen and used MDX to generate pages or content items.
But it sounds like you are interested in using it more a generic component builder? (Is that accurate?)

The content in my app consists of `.mdx` files which each represent a math problem. Each of these math problems is "dynamic" in the sense that you can pass in props to change the specific numbers, symbols, etc for each user when they view the problem.

e.g.

```mdx
What is {props.a} + {props.b}?
```

That's an extremely trivial example, a real problem with a full solution will likely reference those props 10's to 100's of times. Therefore it's necessary to "trim the fat" in terms of syntax.

> first option - a get prop utility

This _could_ work but it's fairly verbose, would add a lot of visual noise, and would require a lot of extra typing.

>second option - nullish coalesing

Unfortunately this won't really work either given how many times they're used.

> ESAST plugins allow the final javascript generated to be modified via an AST.
https://unifiedjs.com/learn/ provides an into to ASTs.
And https://github.com/remcohaszing/recma-nextjs-static-props provides an example of working with props in ESAST.

Since I asked this I _was_ able to get a Recma / ESAST plugin working that works like my proposal above with the magic `getDefaultProps` function and it seems to work pretty well

> My initial reaction/thought is that having a magic function to handle this specific use case sounds like it could cause as much if not more confusion, than it would solve a problem and add clarity.
But open to hearing thoughts otherwise.

Yeah I get you about the magic-ness of it, however this definitely is not the first time I've found MDX's lack of a "function body" limiting.

That leads me to another idea: what if we just renamed the function to be a hook like `useProps`? This would allow people to do pretty much anything they can in normal react component body, including using hooks, setting default props, etc in a fairly clean MDX-like way.

Could work like

```mdx
# My Heading

My prop: {props.a}

My hook prop: {props.fancyProp}

export function useProps(props) {
  const { fancyProp }  = useFancyHook()
  return { ...props, a: 'value', fancyProp }
}
```

 Thoughts?

**Reply by wooorm** [reply]
I don’t see why this has to be done *inside* MDX. You can do this outside of MDX just fine.
You can export your `getDefaultProps` function from the MDX file. And then, in the place where you use MDX components, call `MDXContent(getDefaultProps())`.

---

> however this definitely is not the first time I've found MDX's lack of a "function body" limiting.

I have no clue why functions in JavaScript are limiting. Perhaps you can explain why you think that.

**Reply by mattvague** [reply]
>I don’t see why this has to be done inside MDX. You can do this outside of MDX just fine.
You can export your getDefaultProps function from the MDX file. And then, in the place where you use MDX components, call MDXContent(getDefaultProps()).

Hmm, yeah that would definitely be an option. I feel like it would be a bit tedious though if every component that calls one of my MDX files has to also know to import and use the `getDefaultProps` fn, not to mention adding quite a bit of visual noise. Consider having to read and write `<MyMDX {...getDefaultProps()} />` vs `<MyMDX />` many times in a single doc.

>I have no clue why functions in JavaScript are limiting. Perhaps you can explain why you think that.

Sorry, I feel like that was poorly worded. What I was trying to say is that I've run into a fair number of cases where I wanted to (among other things) be able to assign a value to a local variable that is dependent on other props, should otherwise only be called once, or require a hook. For example:

```jsx
export function MyComponent(props) {
  const propWithDefault = props.a || 'default value'
  const conditionalProp = props.someFlag ? 'option 1' : 'option 2'
  const expensiveProp = someExpensiveComputationThatShouldOnlyHappenOnce()
  const hookProp = useSomeHook()
  
  return (
    <div>
      Some prop w/ default: {propWithDefault}
      Some conditional prop: {conditionalProp}
      Some expensive prop: {expensiveProp}
      Some hook prop: {hookProp}
    </div>
  )
}
```

As far as I understand, there's currently no straightforward way to do any of these in MDX. While I'm sure that there are possible workarounds for all of these cases, none seem like they would be as clean and simple as just being able to intercept and modify the props passed to an MDX component in a JS function of some sort like so

```mdx
export function useProps(props) {
  const propWithDefault = props.a || 'default value'
  const conditionalProp = props.someFlag ? 'option 1' : 'option 2'
  const expensiveProp = someExpensiveComputationThatShouldOnlyHappenOnce()
  const hookProp = useSomeHook()
  
  return { propWithDefault, conditionalProp, expensiveProp, hookProp }
}

Some prop w/ default: {props.propWithDefault}
Some conditional prop: {props.conditionalProp}
Some expensive prop: {props.expensiveProp}
Some hook prop: {props.hookProp}
```

However, all I can speak for is my use case. It's very possible that the majority of MDX users wouldn't have this problem and so following through with my proposal adds needless complexity. If you're still leaning towards the latter, but you think there's some subset of other MDX users that would benefit from this, I could potentially extract it into a separate lib at some point.

Thanks for the thoughtful and timely responses ❤️

**Reply by wooorm** [reply]
> I feel like it would be a bit tedious though if every component that calls one of my MDX files has to also know to import and use the getDefaultProps fn,

I dunno, I don’t really get it 🤷‍♂️ If you’re using `MDXContent` in 100 places, perhaps that’s something you can wrap into some more code, to always call this function.

---

Thanks for the explanation! I still do not see a reason why writing the code you want, and calling and applying it, isn’t enough:

```js
import * as mod from './path/to/file.mdx'
const {default: MDXContent, useProps} = mod
const normalProps = {/*…*/}
const props = useProps ? useProps(normalProps) : normalProps

MDXContent(props)
```

Coming back to your proposal, automatically calling `props = useProps(props)` inside `MDXContent`, perhaps this can first be a plugin, where it can be tested and experimented with, and if it is something that’s very popular, it might later be integrated into MDX? Also sounds like something frameworks typically do for you with things like `getStaticProps` and such instead of something that is MDX-specific

**Reply by mattvague** [reply]
>perhaps that’s something you can wrap into some more code, to always call this function.

Not totally sure what you mean by "always call this function". Do you mean the `useProps`/`getDefaultProps` function? 

If so, then that pretty much is my goal here. It's very possible I'm missing something obvious, but I currently don't see a way to make this all totally automatic without some sort of Recma plugin that modifies either the importing or exporting component to always use `useProps`.

>Thanks for the explanation! I still do not see a reason why writing the code you want, and calling and applying it, isn’t enough:

Where are you proposing that this code would go? Inside of the component that is rendering `MDXContent`? 

>Coming back to your proposal, automatically calling props = useProps(props) inside MDXContent, perhaps this can first be a plugin, where it can be tested and experimented with, and if it is something that’s very popular, it might later be integrated into MDX? Also sounds like something frameworks typically do for you with things like getStaticProps and such instead of something that is MDX-specific

Sounds good! I think I'll probably try it out a bit in my app first and if it seems to work well then I'll extract it into a plugin.

**Reply by mattvague** [reply]
@wooorm @ChristianMurphy Would you prefer that I mark this as answered for now, or leave it as-is until I extract the plugin or figure out some other solution?

**Reply by wooorm** [reply]
How are you importing mdx?

**Reply by mattvague** [reply]
A (very very simple) example MDX problem with MDX imports might look like look like this:

```mdx
import Roots from "~/content/Definitions/Algebra/Roots.mdx";
import Polynomial from "~/content/Definitions/Algebra/Polynomial.mdx";

Find the <Roots/> of the following <Polynomial/>:

$$
  x^2 - 5x +6
$$
```

where `<Roots/>` and `<Polynomial/>` render the word "roots", and "polynomial" respectively and then show a definition on hover. When we end up using a definition in a lot of different files we do usually end up passing it to `MDXProvider` instead to avoid having to import everywhere (I anticipate you suggesting this 😆)

These definition tooltips often have props that (among other things) control which symbols are displayed, and which we want to have default values. For example:

An example of `Roots.mdx` content is:

```mdx
The <dfn>root(s)</dfn> of an equation $\js{props.y}(\js{props.x}) = \js{props.fn}(\js{props.x})$ are
the values of $\js{props.x}$ for which the equality is true.

<Latin.Eg /> the roots of $\js{props.x}^2 = \js{props.root**2}$ are $\js{props.x} = \js{props.root}$ and $\js{props.x} = -\js{props.root}$
because these are the only values for which the expression is true.
```

**Reply by wooorm** [reply]
I was specifically wondering how you *import* MDX, as in, whether you are using Next.js or so. Next.js “hides” much of MDX by default, but it can be changes by defining an `_app.js` or so file.

A solution that I can imagine for this: just like with components which can be received as props but also injects a `_provideComponents()` as can be seen [here](https://mdxjs.com/packages/mdx/#optionsproviderimportsource) (click “Example” to see the diff) when a provider is used, perhaps we could also have a similar provider API for `useMDXProps`? 🤔

Alternatively, React has `defaultProps` with `prop-types`. Potentially something like that can be supported. However, as that doesn’t translate to Vue or other random JSX implementations, I’d need a bunch of convincing?

**Reply by mattvague** [reply]
Very sorry for the delayed response! Been a busy few weeks.

> I was specifically wondering how you import MDX, as in, whether you are using Next.js or so. Next.js “hides” much of MDX by default, but it can be changes by defining an _app.js or so file.

Ahh ok. No, not using Next.js. Currently have a home-rolled setup using webpack with a custom MDXLoader using `@mdx-js/mdx` directly so I have direct access to change whatever I need to.

>A solution that I can imagine for this: just like with components which can be received as props but also injects a _provideComponents() as can be seen [here](https://mdxjs.com/packages/mdx/#optionsproviderimportsource) (click “Example” to see the diff) when a provider is used, perhaps we could also have a similar provider API for useMDXProps? 🤔

Hmmmm, interesting. So it would then just look for a `useMDXProps` hook exported in your MDX file? What would the option be called?

>Alternatively, React has defaultProps with prop-types. Potentially something like that can be supported. However, as that doesn’t translate to Vue or other random JSX implementations, I’d need a bunch of convincing?

Right, thought about this option but had similar reservations.

**Reply by wooorm** [reply]
I was thinking of `useMDXProps(props)` on your (user) side, and injecting `useMDXProps()` on the compiler side

**Reply by mattvague** [reply]
>I was thinking of useMDXProps(props) on your (user) side

Sorry, not sure I'm quite following. Are you proposing that (on the user side, in your MDX file) it would work something like what I've described above like:

```mdx
My Prop: {props.myProp}

export function useMDXProps(props) {
  return { ...props, myProp: 2 }
}
```

Or something else? 

>, and injecting useMDXProps() on the compiler side

And by injecting, are you meaning that the compiled JS would look something like this?

```diff
- export default function MDXContent(props = {}) {
+ export default function MDXContent(initialProps = {}) {
+  const props = useMDXProps(initialProps)

   return MDXLayout
     ? _jsx(MDXLayout, Object.assign({}, props, {children: _jsx(_createMdxContent, {})}))
     : _createMdxContent()
```

**Reply by wooorm** [reply]
Sort of. The providers, such as `@mdx-js/react`, have a `useMDXComponents` method that you as a user can use *outside* of an MDX file, to pass things to be used *inside* that MDX file: https://mdxjs.com/packages/react/#usemdxcomponentscomponents.
That same function can be called, without arguments, to access to current things, from within an MDX file.
With `providerImportSource`, code is generated to do just that.
I am proposing for a) this method `useMdxProps` to be added to the three providers we provide (hah), and b) for the compiler (`@mdx-js/mdx`) to, when `options.providerImportSource` is passed, change:

```diff
-export default function MDXContent(props) {
+export default function MDXContent(_props) {
+  const props = _useMdxProps(_props)
```

**Reply by mattvague** [reply]
Ohhhhh, ok I guess never quite understood how `useMDXComponents` worked then 😆 That's pretty neat.

**Reply by mattvague** [reply]
So, what are the next steps then? Or do you want to let this one simmer?

**Reply by wooorm** [reply]
Next step IMO would be a PR. I think it should work. Want to work on it?

**Reply by mattvague** [reply]
Hey! Yeah would be happy to. Week days are pretty packed but will try to work on it on the weekend a bit

**Reply by wooorm** [reply]
Same, I’m also busy.

This is a bit tough, not for beginners, so here’s some tips on what I’d suggest:

- Create an issue detailing the planned approach, mentioning that you plan to work on it
- Perhaps you can split the work on a) the compiler, b) the react/preact/vue packages into separate PRs.
- One caveat is that provided props could *also* have components, but the provided components should have precedence. Perhaps you want to prefer one or the other, or when providing components, set them on props, or something else?
